lib/introspect.js

```diff
@@ -20,32 +20,68 @@ module.exports = function(obj, callback) {
-                          // TODO: share signal mangling code
-                          var signalFullName = bus.mangle(obj.name, ifName, signame); 
-                          //console.log('trying to listen' +  signalFullName, obj);
-                          bus.signals.on(signalFullName, function(messageBody)
-                          {
+              (function() {
+                  var callbacks = [];
+                  var sigHandlers = [];
+                  function getSigHandler(callback) {
+                      var index;
+                      if ((index = callbacks.indexOf(callback)) == -1) {
+                          index = callbacks.push(callback) - 1;
+                          sigHandlers[index] = function(messageBody) {
                                callback.apply(null, messageBody)
-                          });
-                      });
+                          }
+                      }
+                      return sigHandlers[index];
+                  }
+
+                  function getMatchRule(objName, ifName, signame) {
+                      return "type='signal',path='" + objName + "',interface='" + ifName + "',member='" + signame + "'";
                   }
-              })(ifaceName);
 
-              currentIface.removeListener = function() {
-                console.warn("removeListener not implemented for dbus signals");
-              };
-            
+                  currentIface.addListener = currentIface.on = (function(ifName) {
+                      return function(signame, callback) {
+                          // http://dbus.freedesktop.org/doc/api/html/group__DBusBus.html#ga4eb6401ba014da3dbe3dc4e2a8e5b3ef
+                          // An example is "type='signal',sender='org.freedesktop.DBus', interface='org.freedesktop.DBus',member='Foo', path='/bar/foo',destination=':452345.34'" ...
+                          //var match = "type='signal',sender='"+obj.service.name+ ...
+
+                          var signalFullName = bus.mangle(obj.name, ifName, signame);
+
+                          if (!bus.signals.listeners(signalFullName).length) {
+                              // This is the first time, so call addMatch
+                              var match = getMatchRule(obj.name, ifName, signame);
+                              bus.addMatch(match, function(err, result) {
+                                  if (err) throw new Error(err);
+
+                                  bus.signals.on(signalFullName, getSigHandler(callback));
+                              });
+                          }
+                          else {
+                              // The match is already there, just add event listener
+                              bus.signals.on(signalFullName, getSigHandler(callback));
+                          }
+                      }
+                  })(ifaceName);
+
+                  currentIface.removeListener = (function(ifName) {
```

Comment:
Node's eventemitter does not have off - only https://nodejs.org/api/events.html#events_emitter_removelistener_event_listener